KafkaJS is not re-auhtenticating to all brokers

**Describe the bug**
When I run kafkajs with multiple brokers I see in logs that it authenticated to all of them successfuly
```
[KafkaJS] [SASLOAuthBearerAuthenticator][DEBUG] - Authenticate with SASL OAUTHBEARER kafka-1.flucivja.local:9094
[KafkaJS] [SASLOAuthBearerAuthenticator][DEBUG] - SASL OAUTHBEARER authentication successful kafka-1.flucivja.local:9094

[KafkaJS] [SASLOAuthBearerAuthenticator][DEBUG] - Authenticate with SASL OAUTHBEARER kafka-2.flucivja.local:9094
[KafkaJS] [SASLOAuthBearerAuthenticator][DEBUG] - SASL OAUTHBEARER authentication successful kafka-2.flucivja.local:9094
```

then I see it succesfully receving topics and when `reauthenticationThreshold ` is hit then re-authentication happes but only to first broker
```
[KafkaJS] [SASLOAuthBearerAuthenticator][DEBUG] - Authenticate with SASL OAUTHBEARER kafka-1.flucivja.local:9094
[KafkaJS] [SASLOAuthBearerAuthenticator][DEBUG] - SASL OAUTHBEARER authentication successful kafka-1.flucivja.local:9094
```

then it is working for a while and suddenly I got this error where it is clearly visible that connection was closed by second broker because client didn't re-authenticate to it.
```
[KafkaJS] [Connection][DEBUG] - Kafka server has closed connection kafka-2.flucivja.local:9094
[KafkaJS] [Connection][DEBUG] - disconnecting...
[KafkaJS] [Connection][DEBUG] - disconnected
[KafkaJS] [Runner][DEBUG] - Error while fetching data, trying again...
[KafkaJS] [ConsumerGroup][DEBUG] - Fetching from 1 partitions for 1 out of 1 topics
 [KafkaJS] [Connection][DEBUG] - Request Fetch(key: 1, version: 11)
 [KafkaJS] [Connection][DEBUG] - Response Fetch(key: 1, version: 11)
 [KafkaJS] [Runner][DEBUG] - Error while fetching data, trying again...
... repeated several times
[KafkaJS] [Consumer][ERROR] - Crash: KafkaJSNumberOfRetriesExceeded: Not connected
[KafkaJS] [Runner][DEBUG] - stop consumer group
[KafkaJS] [Consumer][INFO] - Stopped
[KafkaJS] [Consumer][DEBUG] - consumer has stopped, disconnecting
[KafkaJS] [Connection][DEBUG] - disconnecting...
[KafkaJS] [Connection][DEBUG] - disconnected
[KafkaJS] [Consumer][ERROR] - Restarting the consumer
...then consumer is restarted and same happens every 5 minutes
```

**To Reproduce**
You need to run KafkaJS with multiple brokers with SASL OAUTHBEARER auth and server brokers must be configured that client needs to reauthenticate every e.g. 5 minutes.

**Expected behavior**
KafkaJS re-authenticates to all connected brokers

**Environment:**
 - Mac OS and also Linux
 - KafkaJS version 1.15.0
 - NodeJS version 14

--- a/docker-compose.2_4_oauthbearer.yml
+++ b/docker-compose.2_4_oauthbearer.yml
@@ -49,6 +49,7 @@ services:
       # https://github.com/confluentinc/cp-docker-images/blob/master/debian/kafka/include/etc/confluent/docker/log4j.properties.template
       KAFKA_LOG4J_LOGGERS: 'kafka.controller=INFO,kafka.producer.async.DefaultEventHandler=INFO,state.change.logger=INFO'
       CONFLUENT_SUPPORT_METRICS_ENABLE: 'false'
+      KAFKA_CONNECTIONS_MAX_REAUTH_MS: 15000
     volumes:
       - ./testHelpers/certs/kafka.server.keystore.jks:/etc/kafka/secrets/kafka.server.keystore.jks:ro,z
       - ./testHelpers/certs/kafka.server.truststore.jks:/etc/kafka/secrets/kafka.server.truststore.jks:ro,z
@@ -93,6 +94,7 @@ services:
       # https://github.com/confluentinc/cp-docker-images/blob/master/debian/kafka/include/etc/confluent/docker/log4j.properties.template
       KAFKA_LOG4J_LOGGERS: 'kafka.controller=INFO,kafka.producer.async.DefaultEventHandler=INFO,state.change.logger=INFO'
       CONFLUENT_SUPPORT_METRICS_ENABLE: 'false'
+      KAFKA_CONNECTIONS_MAX_REAUTH_MS: 15000
     volumes:
       - ./testHelpers/certs/kafka.server.keystore.jks:/etc/kafka/secrets/kafka.server.keystore.jks:ro,z
       - ./testHelpers/certs/kafka.server.truststore.jks:/etc/kafka/secrets/kafka.server.truststore.jks:ro,z
@@ -137,6 +139,7 @@ services:
       # https://github.com/confluentinc/cp-docker-images/blob/master/debian/kafka/include/etc/confluent/docker/log4j.properties.template
       KAFKA_LOG4J_LOGGERS: 'kafka.controller=INFO,kafka.producer.async.DefaultEventHandler=INFO,state.change.logger=INFO'
       CONFLUENT_SUPPORT_METRICS_ENABLE: 'false'
+      KAFKA_CONNECTIONS_MAX_REAUTH_MS: 15000
     volumes:
       - ./testHelpers/certs/kafka.server.keystore.jks:/etc/kafka/secrets/kafka.server.keystore.jks:ro,z
       - ./testHelpers/certs/kafka.server.truststore.jks:/etc/kafka/secrets/kafka.server.truststore.jks:ro,z

--- a/package.json
+++ b/package.json
@@ -36,7 +36,7 @@
     "test:group:producer": "yarn jest --forceExit --testPathPattern 'src/producer/.*'",
     "test:group:consumer": "yarn jest --forceExit --testPathPattern 'src/consumer/.*.spec.js'",
     "test:group:others": "yarn jest --forceExit --testPathPattern 'src/(?!(broker|admin|producer|consumer)/).*'",
-    "test:group:oauthbearer": "OAUTHBEARER_ENABLED=1 yarn jest --forceExit src/producer/index.spec.js src/broker/__tests__/connect.spec.js src/consumer/__tests__/connection.spec.js src/broker/__tests__/disconnect.spec.js src/admin/__tests__/connection.spec.js",
+    "test:group:oauthbearer": "OAUTHBEARER_ENABLED=1 yarn jest --forceExit src/producer/index.spec.js src/broker/__tests__/connect.spec.js src/consumer/__tests__/connection.spec.js src/broker/__tests__/disconnect.spec.js src/admin/__tests__/connection.spec.js src/broker/__tests__/reauthenticate.spec.js",
     "test:group:broker:ci": "JEST_JUNIT_OUTPUT_NAME=test-report.xml ./scripts/testWithKafka.sh \"yarn test:group:broker --ci --maxWorkers=4 --no-watchman\"",
     "test:group:admin:ci": "JEST_JUNIT_OUTPUT_NAME=test-report.xml ./scripts/testWithKafka.sh \"yarn test:group:admin --ci --maxWorkers=4 --no-watchman\"",
     "test:group:producer:ci": "JEST_JUNIT_OUTPUT_NAME=test-report.xml ./scripts/testWithKafka.sh \"yarn test:group:producer --ci --maxWorkers=4 --no-watchman\"",

--- a/src/broker/__tests__/fetchPartitionAllocation.spec.js
+++ b/src/broker/__tests__/fetchPartitionAllocation.spec.js
@@ -65,7 +65,7 @@ describe('Broker > Fetch', () => {
       flattened[2],
       flattened[4],
     ])
-    fetchRequestMock = jest.fn()
+    fetchRequestMock = jest.fn().mockReturnValue({ request: {} })
 
     await broker.fetch({ maxWaitTime, minBytes, maxBytes, topics })
 

--- a/src/broker/__tests__/reauthenticate.spec.js
+++ b/src/broker/__tests__/reauthenticate.spec.js
@@ -1,0 +1,118 @@
+const {
+  saslOAuthBearerConnectionOpts,
+  newLogger,
+  describeIfOauthbearerEnabled,
+  createConnection,
+} = require('testHelpers')
+
+const { SaslAuthenticate: apiSaslAuthenticateKey } = require('../../protocol/requests/apiKeys')
+const Broker = require('../index')
+
+/**
+ * Session lifetime on broker is configured in docker-compose.2_4_oauthbearer.yml with 15000ms in env KAFKA_CONNECTIONS_MAX_REAUTH_MS
+ * see https://docs.confluent.io/platform/current/installation/configuration/broker-configs.html#brokerconfigs_connections.max.reauth.ms
+ */
+describeIfOauthbearerEnabled('Brokers re-authentication with SASL OAUTHBEARER', () => {
+  let broker
+
+  const wait = (time = 1000) =>
+    new Promise(resolve => {
+      setTimeout(resolve, time)
+    })
+
+  const getAuthCalls = spy =>
+    spy.mock.calls.filter(mock => mock[0].request.apiKey === apiSaslAuthenticateKey)
+
+  test('Does not re-authenticate when still within session validity threshold', async () => {
+    const connection = createConnection(saslOAuthBearerConnectionOpts())
+    const spy = jest.spyOn(connection, 'send')
+    broker = new Broker({
+      connection,
+      logger: newLogger(),
+      reauthenticationThreshold: 2000,
+    })
+    await broker.connect()
+    await wait(1000)
+    await broker.listGroups()
+    await wait(1000)
+    await broker.listGroups()
+    await wait(1000)
+    await broker.listGroups()
+
+    expect(getAuthCalls(spy).length).toBe(1)
+  })
+
+  test('Re-authenticate if needed before making a request', async () => {
+    const connection = createConnection(saslOAuthBearerConnectionOpts())
+    const spy = jest.spyOn(connection, 'send')
+    broker = new Broker({
+      connection,
+      logger: newLogger(),
+      reauthenticationThreshold: 2000,
+    })
+    await broker.connect()
+    await wait(13000)
+    await broker.listGroups()
+
+    expect(getAuthCalls(spy).length).toBe(2)
+  })
+
+  test('Re-authenticate only once with multiple parallel requests', async () => {
+    const connection = createConnection(saslOAuthBearerConnectionOpts())
+    const spy = jest.spyOn(connection, 'send')
+    broker = new Broker({
+      connection,
+      logger: newLogger(),
+      reauthenticationThreshold: 2000,
+    })
+    await broker.connect()
+    await wait(13000)
+    await Promise.all([
+      broker.listGroups(),
+      broker.listGroups(),
+      broker.listGroups(),
+      broker.listGroups(),
+    ])
+
+    expect(getAuthCalls(spy).length).toBe(2)
+  })
+
+  test('Re-authenticate every request sent when re-authentication threshold is same as session lifetime', async () => {
+    const connection = createConnection(saslOAuthBearerConnectionOpts())
+    const spy = jest.spyOn(connection, 'send')
+    broker = new Broker({
+      connection,
+      logger: newLogger(),
+      reauthenticationThreshold: 15000,
+    })
+    await broker.connect()
+    await wait(1000)
+    await broker.listGroups()
+    await wait(1000)
+    await broker.listGroups()
+    await wait(1000)
+    await broker.listGroups()
+
+    expect(getAuthCalls(spy).length).toBe(4)
+  })
+
+  test('Re-authenticates only once with multiple requests sent in parallel when re-authentication threshold is same as session lifetime', async () => {
+    const connection = createConnection(saslOAuthBearerConnectionOpts())
+    const spy = jest.spyOn(connection, 'send')
+    broker = new Broker({
+      connection,
+      logger: newLogger(),
+      reauthenticationThreshold: 15000,
+    })
+    await broker.connect()
+    await wait(1000)
+    await Promise.all([
+      broker.listGroups(),
+      broker.listGroups(),
+      broker.listGroups(),
+      broker.listGroups(),
+    ])
+
+    expect(getAuthCalls(spy).length).toBe(2)
+  })
+})

--- a/src/broker/__tests__/reauthenticate.spec.js
+++ b/src/broker/__tests__/reauthenticate.spec.js
@@ -48,10 +48,10 @@ describeIfOauthbearerEnabled('Brokers re-authentication with SASL OAUTHBEARER', 
     broker = new Broker({
       connection,
       logger: newLogger(),
-      reauthenticationThreshold: 2000,
+      reauthenticationThreshold: 14900,
     })
     await broker.connect()
-    await wait(13000)
+    await wait(1000)
     await broker.listGroups()
 
     expect(getAuthCalls(spy).length).toBe(2)
@@ -63,10 +63,10 @@ describeIfOauthbearerEnabled('Brokers re-authentication with SASL OAUTHBEARER', 
     broker = new Broker({
       connection,
       logger: newLogger(),
-      reauthenticationThreshold: 2000,
+      reauthenticationThreshold: 14900,
     })
     await broker.connect()
-    await wait(13000)
+    await wait(1000)
     await Promise.all([
       broker.listGroups(),
       broker.listGroups(),

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -951,10 +951,10 @@ module.exports = class Broker {
    * @private
    */
   async [PRIVATE.SEND_REQUEST](protocolRequest) {
+    if (!this.isAuthenticated() && isAuthenticatedRequest(protocolRequest.request)) {
+      await this[PRIVATE.AUTHENTICATE]()
+    }
     try {
-      if (!this.isAuthenticated() && isAuthenticatedRequest(protocolRequest.request)) {
-        await this[PRIVATE.AUTHENTICATE]()
-      }
       return await this.connection.send(protocolRequest)
     } catch (e) {
       if (e.name === 'KafkaJSConnectionClosedError') {

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -73,19 +73,13 @@ module.exports = class Broker {
     this.sessionLifetime = Long.ZERO
 
     // The lock timeout has twice the connectionTimeout because the same timeout is used
-    // for the first apiVersions call and it has additional little buffer to last little longer
-    // so when {this.authLock} timeouts within connection, the auth timeout error will be created, not generic lock timeout.
-    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout + 50
+    // for the first apiVersions call
+    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout
     this.brokerAddress = `${this.connection.host}:${this.connection.port}`
 
     this.lock = new Lock({
       timeout: lockTimeout,
       description: `connect to broker ${this.brokerAddress}`,
-    })
-
-    this.authLock = new Lock({
-      timeout: this.authenticationTimeout,
-      description: `authentication to broker ${this.brokerAddress}`,
     })
 
     this.lookupRequest = notInitializedLookup
@@ -95,23 +89,17 @@ module.exports = class Broker {
      * @returns {Promise}
      */
     this[PRIVATE.AUTHENTICATE] = sharedPromiseTo(async () => {
-      try {
-        await this.authLock.acquire()
+      if (this.connection.sasl && !this.isAuthenticated()) {
+        const authenticator = new SASLAuthenticator(
+          this.connection,
+          this.rootLogger,
+          this.versions,
+          this.supportAuthenticationProtocol
+        )
 
-        if (this.connection.sasl && !this.isAuthenticated()) {
-          const authenticator = new SASLAuthenticator(
-            this.connection,
-            this.rootLogger,
-            this.versions,
-            this.supportAuthenticationProtocol
-          )
-
-          await authenticator.authenticate()
-          this.authenticatedAt = process.hrtime()
-          this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
-        }
-      } finally {
-        await this.authLock.release()
+        await authenticator.authenticate()
+        this.authenticatedAt = process.hrtime()
+        this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
       }
     })
   }

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -6,6 +6,7 @@ const { KafkaJSNonRetriableError } = require('../errors')
 const apiKeys = require('../protocol/requests/apiKeys')
 const SASLAuthenticator = require('./saslAuthenticator')
 const shuffle = require('../utils/shuffle')
+const { ApiVersions: apiVersionsApiKey } = require('../protocol/requests/apiKeys')
 
 const PRIVATE = {
   SHOULD_REAUTHENTICATE: Symbol('private:Broker:shouldReauthenticate'),
@@ -22,9 +23,8 @@ const notInitializedLookup = () => {
  * @param request - request from protocol
  * @returns {boolean}
  */
-const isVersionsRequest = request => {
-  const possibleVersionsProtocol = requests.ApiVersions.protocol({ version: request.apiVersion })
-  return possibleVersionsProtocol && possibleVersionsProtocol().request.apiName === request.apiName
+const isAuthenticatedRequest = request => {
+  return request.apiKey !== apiVersionsApiKey
 }
 
 /**
@@ -73,7 +73,7 @@ module.exports = class Broker {
 
     // The lock timeout has twice the connectionTimeout because the same timeout is used
     // for the first apiVersions call
-    const lockTimeout = 2 * this.connection.connectionTimeout
+    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout
     this.brokerAddress = `${this.connection.host}:${this.connection.port}`
 
     this.lock = new Lock({
@@ -951,10 +951,10 @@ module.exports = class Broker {
    * @private
    */
   async [PRIVATE.SEND_REQUEST](protocolRequest) {
-    if (!this.isAuthenticated() && !isVersionsRequest(protocolRequest.request)) {
-      await this[PRIVATE.AUTHENTICATE]()
-    }
     try {
+      if (!this.isAuthenticated() && isAuthenticatedRequest(protocolRequest.request)) {
+        await this[PRIVATE.AUTHENTICATE]()
+      }
       return await this.connection.send(protocolRequest)
     } catch (e) {
       if (e.name === 'KafkaJSConnectionClosedError') {

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -7,6 +7,7 @@ const apiKeys = require('../protocol/requests/apiKeys')
 const SASLAuthenticator = require('./saslAuthenticator')
 const shuffle = require('../utils/shuffle')
 const { ApiVersions: apiVersionsApiKey } = require('../protocol/requests/apiKeys')
+const sharedPromiseTo = require('../utils/sharedPromiseTo')
 
 const PRIVATE = {
   SHOULD_REAUTHENTICATE: Symbol('private:Broker:shouldReauthenticate'),
@@ -88,6 +89,31 @@ module.exports = class Broker {
     })
 
     this.lookupRequest = notInitializedLookup
+
+    /**
+     * @private
+     * @returns {Promise}
+     */
+    this[PRIVATE.AUTHENTICATE] = sharedPromiseTo(async () => {
+      try {
+        await this.authLock.acquire()
+
+        if (this.connection.sasl && !this.isAuthenticated()) {
+          const authenticator = new SASLAuthenticator(
+            this.connection,
+            this.rootLogger,
+            this.versions,
+            this.supportAuthenticationProtocol
+          )
+
+          await authenticator.authenticate()
+          this.authenticatedAt = process.hrtime()
+          this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
+        }
+      } finally {
+        await this.authLock.release()
+      }
+    })
   }
 
   /**
@@ -126,19 +152,7 @@ module.exports = class Broker {
       }
 
       this.lookupRequest = lookup(this.versions)
-      await this[PRIVATE.AUTHENTICATE]()
-    } finally {
-      await this.lock.release()
-    }
-  }
 
-  /**
-   * @private
-   * @returns {Promise}
-   */
-  async [PRIVATE.AUTHENTICATE]() {
-    try {
-      await this.authLock.acquire()
       if (this.supportAuthenticationProtocol === null) {
         try {
           this.lookupRequest(apiKeys.SaslAuthenticate, requests.SaslAuthenticate)
@@ -153,23 +167,9 @@ module.exports = class Broker {
         })
       }
 
-      if (
-        (this.authenticatedAt == null || this[PRIVATE.SHOULD_REAUTHENTICATE]()) &&
-        this.connection.sasl
-      ) {
-        const authenticator = new SASLAuthenticator(
-          this.connection,
-          this.rootLogger,
-          this.versions,
-          this.supportAuthenticationProtocol
-        )
-
-        await authenticator.authenticate()
-        this.authenticatedAt = process.hrtime()
-        this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
-      }
+      await this[PRIVATE.AUTHENTICATE]()
     } finally {
-      await this.authLock.release()
+      await this.lock.release()
     }
   }
 

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -78,10 +78,17 @@ module.exports = class Broker {
    * @public
    * @returns {boolean}
    */
+  isAuthenticated() {
+    return this.authenticatedAt != null && !this[PRIVATE.SHOULD_REAUTHENTICATE]()
+  }
+
+  /**
+   * @public
+   * @returns {boolean}
+   */
   isConnected() {
     const { connected, sasl } = this.connection
-    const isAuthenticated = this.authenticatedAt != null && !this[PRIVATE.SHOULD_REAUTHENTICATE]()
-    return sasl ? connected && isAuthenticated : connected
+    return sasl ? connected && this.isAuthenticated() : connected
   }
 
   /**
@@ -103,35 +110,45 @@ module.exports = class Broker {
       }
 
       this.lookupRequest = lookup(this.versions)
-
-      if (this.supportAuthenticationProtocol === null) {
-        try {
-          this.lookupRequest(apiKeys.SaslAuthenticate, requests.SaslAuthenticate)
-          this.supportAuthenticationProtocol = true
-        } catch (_) {
-          this.supportAuthenticationProtocol = false
-        }
-
-        this.logger.debug(`Verified support for SaslAuthenticate`, {
-          broker: this.brokerAddress,
-          supportAuthenticationProtocol: this.supportAuthenticationProtocol,
-        })
-      }
-
-      if (this.authenticatedAt == null && this.connection.sasl) {
-        const authenticator = new SASLAuthenticator(
-          this.connection,
-          this.rootLogger,
-          this.versions,
-          this.supportAuthenticationProtocol
-        )
-
-        await authenticator.authenticate()
-        this.authenticatedAt = process.hrtime()
-        this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
-      }
+      await this.authenticate()
     } finally {
       await this.lock.release()
+    }
+  }
+
+  /**
+   * @private
+   * @returns {Promise}
+   */
+  async authenticate() {
+    if (this.supportAuthenticationProtocol === null) {
+      try {
+        this.lookupRequest(apiKeys.SaslAuthenticate, requests.SaslAuthenticate)
+        this.supportAuthenticationProtocol = true
+      } catch (_) {
+        this.supportAuthenticationProtocol = false
+      }
+
+      this.logger.debug(`Verified support for SaslAuthenticate`, {
+        broker: this.brokerAddress,
+        supportAuthenticationProtocol: this.supportAuthenticationProtocol,
+      })
+    }
+
+    if (
+      (this.authenticatedAt == null || this[PRIVATE.SHOULD_REAUTHENTICATE]()) &&
+      this.connection.sasl
+    ) {
+      const authenticator = new SASLAuthenticator(
+        this.connection,
+        this.rootLogger,
+        this.versions,
+        this.supportAuthenticationProtocol
+      )
+
+      await authenticator.authenticate()
+      this.authenticatedAt = process.hrtime()
+      this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
     }
   }
 
@@ -361,6 +378,9 @@ module.exports = class Broker {
    * @returns {Promise}
    */
   async heartbeat({ groupId, groupGenerationId, memberId }) {
+    if (!this.isAuthenticated()) {
+      await this.authenticate()
+    }
     const heartbeat = this.lookupRequest(apiKeys.Heartbeat, requests.Heartbeat)
     return await this[PRIVATE.SEND_REQUEST](heartbeat({ groupId, groupGenerationId, memberId }))
   }

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -72,8 +72,9 @@ module.exports = class Broker {
     this.sessionLifetime = Long.ZERO
 
     // The lock timeout has twice the connectionTimeout because the same timeout is used
-    // for the first apiVersions call
-    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout
+    // for the first apiVersions call and it has additional little buffer to last little longer
+    // so when {this.authLock} timeouts within connection, the auth timeout error will be created, not generic lock timeout.
+    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout + 50
     this.brokerAddress = `${this.connection.host}:${this.connection.port}`
 
     this.lock = new Lock({

--- a/src/broker/index.js
+++ b/src/broker/index.js
@@ -10,11 +10,21 @@ const shuffle = require('../utils/shuffle')
 const PRIVATE = {
   SHOULD_REAUTHENTICATE: Symbol('private:Broker:shouldReauthenticate'),
   SEND_REQUEST: Symbol('private:Broker:sendRequest'),
+  AUTHENTICATE: Symbol('private:Broker:authenticate'),
 }
 
 /** @type {import("../protocol/requests").Lookup} */
 const notInitializedLookup = () => {
   throw new Error('Broker not connected')
+}
+
+/**
+ * @param request - request from protocol
+ * @returns {boolean}
+ */
+const isVersionsRequest = request => {
+  const possibleVersionsProtocol = requests.ApiVersions.protocol({ version: request.apiVersion })
+  return possibleVersionsProtocol && possibleVersionsProtocol().request.apiName === request.apiName
 }
 
 /**
@@ -63,12 +73,17 @@ module.exports = class Broker {
 
     // The lock timeout has twice the connectionTimeout because the same timeout is used
     // for the first apiVersions call
-    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout
+    const lockTimeout = 2 * this.connection.connectionTimeout
     this.brokerAddress = `${this.connection.host}:${this.connection.port}`
 
     this.lock = new Lock({
       timeout: lockTimeout,
       description: `connect to broker ${this.brokerAddress}`,
+    })
+
+    this.authLock = new Lock({
+      timeout: this.authenticationTimeout,
+      description: `authentication to broker ${this.brokerAddress}`,
     })
 
     this.lookupRequest = notInitializedLookup
@@ -110,7 +125,7 @@ module.exports = class Broker {
       }
 
       this.lookupRequest = lookup(this.versions)
-      await this.authenticate()
+      await this[PRIVATE.AUTHENTICATE]()
     } finally {
       await this.lock.release()
     }
@@ -120,35 +135,40 @@ module.exports = class Broker {
    * @private
    * @returns {Promise}
    */
-  async authenticate() {
-    if (this.supportAuthenticationProtocol === null) {
-      try {
-        this.lookupRequest(apiKeys.SaslAuthenticate, requests.SaslAuthenticate)
-        this.supportAuthenticationProtocol = true
-      } catch (_) {
-        this.supportAuthenticationProtocol = false
+  async [PRIVATE.AUTHENTICATE]() {
+    try {
+      await this.authLock.acquire()
+      if (this.supportAuthenticationProtocol === null) {
+        try {
+          this.lookupRequest(apiKeys.SaslAuthenticate, requests.SaslAuthenticate)
+          this.supportAuthenticationProtocol = true
+        } catch (_) {
+          this.supportAuthenticationProtocol = false
+        }
+
+        this.logger.debug(`Verified support for SaslAuthenticate`, {
+          broker: this.brokerAddress,
+          supportAuthenticationProtocol: this.supportAuthenticationProtocol,
+        })
       }
 
-      this.logger.debug(`Verified support for SaslAuthenticate`, {
-        broker: this.brokerAddress,
-        supportAuthenticationProtocol: this.supportAuthenticationProtocol,
-      })
-    }
+      if (
+        (this.authenticatedAt == null || this[PRIVATE.SHOULD_REAUTHENTICATE]()) &&
+        this.connection.sasl
+      ) {
+        const authenticator = new SASLAuthenticator(
+          this.connection,
+          this.rootLogger,
+          this.versions,
+          this.supportAuthenticationProtocol
+        )
 
-    if (
-      (this.authenticatedAt == null || this[PRIVATE.SHOULD_REAUTHENTICATE]()) &&
-      this.connection.sasl
-    ) {
-      const authenticator = new SASLAuthenticator(
-        this.connection,
-        this.rootLogger,
-        this.versions,
-        this.supportAuthenticationProtocol
-      )
-
-      await authenticator.authenticate()
-      this.authenticatedAt = process.hrtime()
-      this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
+        await authenticator.authenticate()
+        this.authenticatedAt = process.hrtime()
+        this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
+      }
+    } finally {
+      await this.authLock.release()
     }
   }
 
@@ -378,9 +398,6 @@ module.exports = class Broker {
    * @returns {Promise}
    */
   async heartbeat({ groupId, groupGenerationId, memberId }) {
-    if (!this.isAuthenticated()) {
-      await this.authenticate()
-    }
     const heartbeat = this.lookupRequest(apiKeys.Heartbeat, requests.Heartbeat)
     return await this[PRIVATE.SEND_REQUEST](heartbeat({ groupId, groupGenerationId, memberId }))
   }
@@ -934,6 +951,9 @@ module.exports = class Broker {
    * @private
    */
   async [PRIVATE.SEND_REQUEST](protocolRequest) {
+    if (!this.isAuthenticated() && !isVersionsRequest(protocolRequest.request)) {
+      await this[PRIVATE.AUTHENTICATE]()
+    }
     try {
       return await this.connection.send(protocolRequest)
     } catch (e) {

--- a/testHelpers/index.js
+++ b/testHelpers/index.js
@@ -116,7 +116,7 @@ const saslOAuthBearerConnectionOpts = () =>
     sasl: {
       mechanism: 'oauthbearer',
       oauthBearerProvider: () => {
-        const token = jwt.sign({ sub: 'test' }, 'abc', { algorithm: 'none' })
+        const token = jwt.sign({ sub: 'test' }, undefined, { algorithm: 'none', expiresIn: '1h' })
 
         return {
           value: token,

